### PR TITLE
[RFC][dynamo] Narrow GB6297 to only graph-break on differentiable duplicate tensor inputs

### PR DESCRIPTION
**Pull request by Fiery**
Summary:
D105712843 added GB6297 which graph-breaks whenever the same tensor is passed to multiple forward inputs of `autograd.Function.apply()`. This was a correctness fix for a real gradient-loss bug where Dynamo's HOP deduplication silently drops one backward gradient contribution.

However, the check is overly broad -- it fires on ALL duplicate tensor inputs regardless of differentiability. Non-differentiable tensors (integer dtype, `requires_grad=False`) always receive `None` gradients in backward, so deduplicating them is harmless. The overwrite in `outer_fwd_proxy_to_bwd_node` is `None -> None`, which is idempotent.

This causes a significant regression in production Ads training models (ATLAS) where `cu_seqlens` (an int32 index tensor) is passed as both `cu_seqlens_q` and `cu_seqlens_k` to `FlashAttnVarlenFunc.apply()` in the self-attention path. The graph break cascades: GB6297 inside a `for layer in self.layers` loop triggers GB7000 (graph break in loop), which skips the entire `Transformer.forward` frame. This disables both PT2 compilation and AutoAC activation checkpointing for 88 transformer layers, causing peak GPU memory to spike from 85% to 99%.

This diff narrows the check to only fire when `fake_tensor.requires_grad or fake_tensor.dtype.is_floating_point`. The condition is intentionally conservative:
- `requires_grad=True`: gradient flows through this tensor, deduplication loses a contribution
- `is_floating_point` (even with `requires_grad=False`): could require grad at runtime with different inputs
- Integer/boolean tensors: can never carry gradients, deduplication is always safe

Authored with Claude.

Test Plan:
Existing test `test_duplicate_input_accumulates_grad` continues to pass -- differentiable duplicate (float, `requires_grad=True`) still graph-breaks with `fullgraph=True` and falls back correctly with `fullgraph=False`.

New test `test_duplicate_nondifferentiable_input_no_graph_break` -- integer duplicate tensor compiles with `fullgraph=True` without graph break.

```
buck2 test mode/opt fbcode//caffe2/test/dynamo:test_autograd_function -- -r test_duplicate_input_accumulates_grad
buck2 test mode/opt fbcode//caffe2/test/dynamo:test_autograd_function -- -r test_duplicate_nondifferentiable_input_no_graph_break
```

E2E validation pending: rerun ATLAS training job with this fix to confirm memory returns from 99% to ~85%.

Differential Revision: D106449323

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by zou3519** [comment]
@claude review these changes

**Review by zou3519 — CHANGES_REQUESTED** [review]
tests failing, easycla needs to be fixed, i have not read the code.
